### ISSUE TITLE
Editor ID required but no communication about it

### ISSUE THREAD

**Issue by darobin** [w3c/specberus#144]
Our users are surprised (in this case in https://w3c.github.io/webappsec/specs/csp-pinning/published/2015-02-FPWD.html) that their "Editor" field is failing when it looks perfectly fine. That is because the deployed checker now requires an editor ID to be defined but no one knows about this.

Either this change should be backed out as a requirement (it probably should be for the time being) or it should be well-advertised on spec-prod@ and chairs@ so that people know they have to define it.

**Comment by astorije** [comment]
By requirement, you mean one of [Echidna's current limitations](https://github.com/w3c/echidna/wiki/How-to-use-Echidna#current-limitations)?

@plehegar, can you add this to the public announcement you wanted to make?

**Comment by darobin** [comment]
I'm not sure what you mean by limitations, I don't see a requirement for editor IDs listed there.

The current specberus fails unless at least one editor has their data-editor-id specified; I think that's clearly a bug.

**Comment by astorije** [comment]
Hey @darobin, I was trying to understand what you meant by "should be backed out as a requirement".

There must be editors in the document, and these editors must have the `data-editor-id` set, it is not a bug, it's a feature :-)
I'll make sure @plehegar communicates about it and/or that it's documented somewhere.

**Comment by darobin** [comment]
So, we're talking about transitions in another thread... Transitions suck but they're better than shipping a new constraint without telling anyone! :)

If we're going to enforce `data-editor-id` at the very least people should be made aware of it. I don't even know where to get the ID in the first place! And if it's a requirement, ReSpec needs to error when it's not present, etc.

All I'm saying is that this requires a bit of coordination...

**Comment by astorije** [comment]
There there. We might not have told plenty of people, but you were aware as this was discussed and decided thoroughly during the F2F where you showed up. It's also in a tremendous amount of emails and I thought the endless discussions were over, apparently not.

That being said, if you're unhappy with the communication, please email the right people (you can start with @tguild and @plehegar) rather than slapping the developers. This decision came from upstairs and we implemented it.

The ID is the user ID in `https://www.w3.org/users/[userId]`. We didn't reach a better solution when we discussed it, although I wish we had. I am closing this issue as this won't go anywhere, unless you have a better solution to discuss in a new issue.

**Comment by plehegar** [comment]
I don't think we can close the issue that easily. We have two sets of users for specberus:
- one set uses for echidna
- other set uses it as an alternatve for old pubrules

Here is what I think we could:

remove the error in specberus but still generates the metadata. Add the metadata check in echidna or add a small metadata checker to the echidna jobs.

**Comment by tguild** [comment]
As echidna uses its own instance of specberus perhaps we can configure this to be a warning instead of error on the public one http://www.w3.org/2014/10/pubrules/
That warning can link to our documentation... and serve as communication vehicle.  It should remain an error in the specberus instance used by echidna as it is a requirement.

**Comment by plehegar** [comment]
Sounds good to me. I did document last week the data-editor-id at
 https://github.com/w3c/echidna/wiki/Preparing-your-document

**Comment by astorije** [comment]
I would not introduce the ability to make Specberus differ depending on where it is used. In terms of code, it creates a configuration dependency and therefore the possibility of a bug/error. Not to mention it opens the door to a whole set of things we don't want to, that is every time we are not sure what to do we use this ability.

I would very much prefer to find a global solution that is equally treated whether we are using the standalone checker or as part of the Echidna's dependency. We managed to do it with everything else so far and I am sure this can be done here as well.

**Comment by darobin** [comment]
I'm not slapping anyone here, but Echidna is only one user for Specberus. As @plehegar points out there are also people who need a pubrules checker. Right now, due to this change, people are receiving errors that they don't even know how to fix. It doesn't matter who decided what, how this came to be, how it may help with this or that project: if a validation tool gives an error for a requirement you didn't know existed and can't know how to fix, that's a bug.

I think there are two possible cases here:

### Case 1: Editor ID is generally useful (not just for Echidna but also for other use case) and we plan to make use of it across the board.

In this case then it can make sense to add it as a requirement, assuming there are several compelling use cases. If so, this needs to be announced to spec-prod, there needs to be a grace period while people adjust, etc.

### Case 2: Editor ID is pretty much only useful to Echidna

In this case, it shouldn't be enforced on everyone. That's just yet another hoop to make editors' lives more miserable. If Echidna requires it, then that's opt-in and editors will likely be happy to pay the small extra cost in exchange for autopublication. It doesn't matter if it leads to branching in the code, code should make users' lives easier and not the other way around.

To the best of my knowledge the situation is more like (2), but it's entirely possible that I don't know about all the use cases.

**Comment by astorije** [comment]
> That's just yet another hoop to make editors' lives more miserable.

Definitely! Making people (and specifically) editors' lives more miserable is my only goal when working my a*s off on this project at nights and weekends.

> It doesn't matter if it leads to branching in the code, code should make users' lives easier and not the other way around.

Right, and therefore being unwilling to cross out the dumb solutions (branching in the code? seriously?!) is acceptable.

> All I'm saying is that this requires a bit of coordination...

There was complete coordination on this, and as a matter of fact you were among the first ones made aware of since you were there every time it was discussed.

All of that being said, I also dislike the editor ID solution. I always disliked it but it was always presented as a temporary solution that would at least make the system evolve until it gets more mature, which would be a good time to find a better solution. Now all of a sudden it becomes the one true priority we absolutely need to fix or the consortium will collapse, I have to admit I'm confused by this priority shifting... So, at the time I came up with different solutions to improve that, and they were all crossed out for the editor ID solution because of timing.

Now say you really push on having 2 different Specberus behaviors (I think I could not :-1: more). Let's face it, this will at first be chosen for 1 issue (the editors ID) but soon enough plenty of issues will be solved that way, and you don't want that. I won't be talking about the maintenance nightmare for the developers. The result for the user will be that their documents will pass Specberus but fail the automated publication system with a report stating they don't pass Specberus. It doesn't sound like a life made easier to me.

Nightmare for the users and the developers, that's pretty much what I am trying to do by crossing out hackish solutions. Good thing is that we have plenty of others:

1. Have a new type of errors for Specberus, `EchidnaErrors`,  similar as `Error`, `Warning`, ... If the only thing you want is a green badge, then the result could say "Your document is Specberus compliant but be careful as it will fail the automated publication system". Disadvantage is that it "ties" the projects together, but only in wording.

2. Admit that having an error for the editors when the ID is not set is not a big deal and close this discussion. After all, the long term goal of this Echidna is that **all** documents will be published through it, requiring manual intervention on certain types of documents. The key here is one system for everything, and therefore uniformity. Note that document uniformity is also an argument on its own...

3. Find a solution to extract the editors without editor IDs. That would be an interesting project (machine learning \o/) but time consuming and with a lot of edge cases. Probably not what we want in terms of timing.

4. Have a new checker in Echidna's process as @plehegar suggested when he said "add a small metadata checker to the echidna jobs". Actually, this was discussed before and I think it makes a lot of sense. For example, checking the publication date against today's date is now done in Specberus. I didn't understand why as we agreed it should be in a separate checker before, but I would not be surprised if it was done that way to save time. I believe this is the right solution to adopt. And I'm sure I am not the only one, but note that since it would have taken a bit of time, shortcuts were taken. Now, trying to find a quick solution built on a shortcut (aka hack) is not right. I am all for shortcuts when we have to find a quick solution, but making hacks permanent is not something you want.

Therefore, I think the right solution here is **4**. It allows a nice separate of concerns, and it can made available on its own to check your document before publishing it. Specberus would just return all the metadata that would be checked against the W3C database by the other checker. And with the proper framework, this "metadata checker" can serve many purposes without being a frustration for the users, or a pain for the developers.
That said, we go back to the root of all arguments: timing. If **4** is the ultimate solution, you have to choose between a project that will take more time before being released, or accepting a temporary solution to make it work, that is **2**.
It does not involve any hacks or tie-ins between the projects as **1** would. And finally, the day the checker can identify the users in a smarter way (that's **3**), then documents can omit the IDs, but the good point is that Specberus would not be affected.

Voilà, you have my suggestion as a developer on both systems, user of the old pubrules and the Specberus, and the guy who publishes the documents. Hopefully I explained myself well enough. Let me know if something doesn't make sense and I'll rephrase.

**Comment by astorije** [comment]
I do write long emails, don't I?!

I just wanted to add, because I never suggest solutions without being part of them: suggesting **4** as the solution we want to adopt and having **2** for the time being is not a way to keep you guys silent :-)
I'll work on this additional checker once the current tasks are done, and will move the echidna-specific checks to this checker. I mean, would, if you guys agree. I think this echidna checker was a long-term goal anyway, so I don't see a good reason not to do it (unless I am missing a use case).

**Comment by darobin** [comment]
Man, I don't know what to say. This is not a criticism of you or your work, it's just a bug report. One of the primary complaints from editors is that there are too many things they need to think about that are unrelated to creating a good spec and that they feel arbitrary. Adding a new one, with no prior explanation, and due to reasons that only benefit a subset of editors is just that: a bug. And one we can fix. You seem to think I feel this is the end of the world and you're a horrible person or something. That's not at all what I'm saying! If I thought that were the case I wouldn't bother reporting a bug, I'd just go run my own checker somewhere else. It's just a bug, man…

Also, it's a bug I'd be happy to fix myself; but I need to know what works for Echidna. I knew from the meeting that **Echidna** would need an editor ID, and that's fine. Perhaps not perfect as you say but it's okay. People get a new service, they can spend a little time figuring out how it works.

I don't think that supporting different configurations for different contexts in which Specberus is used is all that bad, we have some already and this could just be a new checkbox "[ ] Ensure document is valid for automated publication (help?)". One reason why I don't think it'll break the system is because the *whole* profile/rules/options architecture is *built* on the knowledge that the publication rules are largely arbitrary, all over the place, and full of exceptions and alternatives. Adding a new option doesn't even register compared to the chaos that's already handled :) But if you think that's too bad I don't mind, I just want a solution and hey, it's your baby.

I think we can rule out (3) machine learning for now. Going for (2) makes Specberus needlessly painful to use for people who aren't in the Echidna system. By necessity, that's quite a few people. I think it would be really nice to make it possible for them to use Specberus instead of the old pubrules.

I can live with (1) or (4). Having said that, I could have implemented the checkbox option in less time than it took me to write this comment :)

:kiss:

**Comment by astorije** [comment]
Fair enough, sorry about that :-)

I see what you are suggesting now. I also see the error now showing up on a different document.

My main concern was that having the code diverging would be a source of errors for sure. Now simply having an extra checkbox on the UI changes the deal (I didn't think of that :p) as you can now replicate the output.

- **3** can be ruled out indeed. I'd love to be the guy doing it, but I'm not sure it will ever happen :D
- **1** would probably work well with your suggestion, as the checkbox would treat an `EchidnaError` as an `Error` if checked, and as a `Warning` otherwise. But I still think it's gonna be a pain to maintain.
- **2**... I can see why you don't want that. We would otherwise be publishing documents with an `Error` inside and editors would ask every now and then why it's not compliant. Making it a `Warning` instead (as a temporary temporary solution) would let people publish documents without editors at the moment, but I can come up with a fix during a later step of the publication process.
- **4** is, IMO, still the optimal solution, and here is why: as much as I like the idea of a Specberus checking everything and re-using its nice "checker" framework, some rules will be a pain to implement. For example, @plehegar wanted to check that the WG properly exists (in the W3C DB), at some point we'll want to check that the editor's ID also exists (in the W3C DB), ... My point is that some checks will require interaction with our DB, probably most checks that are publication-related rather than well-formedness related. Now, if you think it's worth having Specberus integrate all of that itself in the near future, then it's fine by me. If not, we'll need an extra checker for the extracted metadata soon anyway. Does that make sense?

**Comment by astorije** [comment]
@darobin and I just had a call to clear this furthermore and he told me he was not bothered by having a Specberus with more powa. That is, when someday we decide to check the validity of the editor IDs, the existence of the WG, etc., all coming from HTTP APIs, they can be checked in Specberus, like what it is done with the validators at the moment.
Which means essentially that we do not need an extra checker for the metadata.

@darobin threw a nice idea, I think, in that instead of having an extra option when calling Specberus (in terms of code, not UI), we can have a different profile, `WD-Echidna` or whatever, that extends the `WD` rules and add the publication-specific rules. We'll have as many new profiles as Echidna support them (so... one at the moment). These profiles will contain at the moment things like checking that the publication date is today's date, the editor's IDs are present, ... A good thing IMO is that I think that most (if not all) of these rules will apply cross-profiles so the Echidna-specific rules can even exist in one file and be used across profiles.
In Echidna, it's then just a matter of calling the right profile.
@darobin also mentioned that the UI can differ for that and, instead of showing all the profiles twice, a checkbox "Automated publication" can do the job and even disable profiles that are not automatically publishable.

So my only concern was how much burden would Specberus soon have, and @darobin told me it was not a big deal. What do you guys think about that?

**Comment by astorije** [comment]
@tripu, just so you know, I am planning to start working on this tomorrow to wrap this up ASAP. I'll open a PR as soon as I have my first commit so that you guys see where I am.

**Comment by tripu** [comment]
@astorije, you guys lost me somewhere around volume XXVIII of your conversation&hellip; I'm getting the gist of it again.

(The reason we opted for Specberus taking care of *both* pubrules-checking and metadata-extraction was that it's simpler and more efficient to do those two things at the same time. Otherwise we would have pretty much the same document-parsing and the same regular expressions in two steps in Echidna: one for metadata-extraction, and then again in Specberus. I think that would be redundant, more difficult to maintain, and Echidna would be slower processing documents. So it's not a hack, IMHO, a good way to do things.)

To make sure I understand what you suggest we do, @astorije: is this it?
* Add new Symfony endpoints that will allow for new checks that are related to the DB (eg, *&ldquo;that the editor's ID also exists in the W3C DB&rdquo;*).
* Keep Specberus as it is now basically, doing validation and metadata extraction.
* Add a new step in Echidna that would query those Symfony endpoints to do these few extra *Echidna-specific* checks.
* Duplicate spec profiles (ie, right now, just add `WD-Echidna` as you say) in Echidna only; not in Specberus.

Is that it?

**Comment by astorije** [comment]
Hey @tripu, sorry about that, that a very confusing topic indeed :-)

> (The reason we opted for Specberus taking care of both pubrules-checking and metadata-extraction [...] So it's not a hack, IMHO, but a good way to do things.)

Oh I totally get that, and that's not what I was calling a hack. The hack, IMO, would be to hard-code things for Echidna only, or to have 2 branches on Specberus for that, but fortunately it is not where we are going.

> - Add new Symfony endpoints that will allow for new checks that are related to the DB (eg, “that the editor's ID also exists in the W3C DB”).

That's most likely what needs to happen, but it would be for step 2. Put it in another way, the current issue here doesn't need this endpoint, but it will be needed pretty soon.

> - Keep Specberus as it is now basically, doing validation and metadata extraction.

Yes, we don't extract (pun intended) metadata extraction from Specberus. I would still like the idea of, in the future, splitting these 2 (to be able to run the token checker upfront, or all checkers in parallel) but I don't see how we would do that. The benefits you mentioned in keeping them altogether are huge and we cannot afford to lose them.

> - Add a new step in Echidna that would query those Symfony endpoints to do these few extra Echidna-specific checks.

Actually no, @darobin's proposal is to make these checks in Specberus. I was against this at first and in favor of having an extra checker in Echidna, but it's a bit of a shame when you consider that the testing framework is already there, in Specberus.

> - Duplicate spec profiles (ie, right now, just add WD-Echidna as you say) in Echidna only; not in Specberus.

Assuming you meant "storing the Echidna-specific Specberus profile within the Echidna repository": that's an idea, but we might benefit from having in Specberus. The main reason is that we don't want people to have Specberus (on the web) to pass and Echidna-flavored Specberus to fail.
One of @darobin's idea was to make it a profile that can be selected either from the profile list, or from a checkbox "Check for automatic publication". That way, if you publish old-style, you don't care about a missing editor-id in your document. If you publish through Echidna, you still have a way to check the document before actually publishing.

Makes sense?
I couldn't work on that today unfortunately, I'll do that next week.

**Comment by astorije** [comment]
Here we go: #153 !

**Comment by astorije** [comment]
#153 was merged, so closing this one.
